dev-docs/RFCs/v5.0/multi-viewport-rfc.md

```diff
@@ -0,0 +1,201 @@
+ RFC - Multi Viewports
+
+* **Author**: Ib Green
+* **Date**: August 10, 2017
+* **Status**: **Draft**
+
+Notes:
+* This RFC started out as an extension of an initial left/right viewport prototype for WebVR integration made by Xiaoji Chen.
+
+
+## Motivation
+
+It is common in 3D applications to render a 3D scene multiple times, with different cameras:
+* To show views from multiple viewpoints (cameras), e.g. in a split screen setup.
+* To show a detail view (e.g, first person), and an overlaid, smaller "map" view (e.g. third person or top down, zoomed out to show where the primary viewpoint is).
+* To support stereoscopic rendering, where left and right views are needed, providing the necessary parallax between left and right eye.
+* For rendering into offscreen framebuffers, which can then be used for e.g. advanced visual effects, screen shot solutions, overlays onto DOM elements outside of the primary deck.gl canvas (e.g. a video).
+
+
+# Problem Statement
+
+* In v4.1, deck.gl creates a canvas and allows only a single viewport that always renders in the top-left corner of that canvas.
+* deck.gl is also limited to rendering into a single canvas, due to the one-to-one relation between a canvas and a `WebGLRenderingContext` (that said it is possible to render into framebuffers that can be extracted).
+In addition
+
+
+## Proposal 1: New `DeckGL.viewports` property
+
+Allows the main `DeckGL` component to accept a list of Viewports (and/or viewport descriptors). This will be in addition to, or preferably instead of, the current `viewport` prop.
+
+Viewports can be side-by-side (top and bottom in this first example). Note how the application controls both the height and the y position of the two viewports.
+```js
+  <DeckGL viewports=[
+    new FirstPersonViewport({...viewprops, height: viewprops.height / 2}),
+    new WebMercatorViewport({...viewprops, y: viewprops.height / 2, height: viewprops.height / 2}),
+    ...
+  ]/>
+```
+
+Side-by-side is of course essential for stereoscopic rendering (and conveniently, the base deck.gl viewport can directly accept view and projection matrices from the WebVR API):
+```js
+  <DeckGL viewports=[
+    // left eye viewport
+    new Viewport({
+      width: viewprops.width / 2,
+      viewMatrix: leftViewMatrix, projectionMatrix: leftProjectionMatrix
+    }),
+    // right eye
+    new Viewport({
+      width: viewprops.width / 2, x: viewprops.width / 2
+      viewMatrix: rightViewMatrix, projectionMatrix: rightProjectionMatrix
+    }),
+    ...
+  ]/>
+```
+
+Or they can overlap, (e.g. having a small overview map in the bottom middle of the screen overlaid over the main view)
```

Comment:
> e.g. having a small overview map in the bottom middle of the screen overlaid over the main view
How does the system decide which viewport should be on top when there are multiple overlapping viewports?